AxonDeepSeg/morphometrics/compute_morphometrics.py

```diff
@@ -589,37 +589,51 @@ def remove_outside_nerve(pred_axon, pred_myelin, pred_nerve):
 
     return pred_axon, pred_myelin
 
+def cleanup_nerve_mask(nerve_mask, min_object_size=5000, max_hole_size=5000):
+    """
+    Removes small objects, fills holes up to max_hole_size, etc.
+    nerve_mask: a binary numpy array where True=inside nerve, False=background
+    """
+    filled = morphology.remove_small_holes(nerve_mask, area_threshold=max_hole_size)
+    cleaned = morphology.remove_small_objects(filled, min_size=min_object_size)
+    
+    return cleaned
+
 def compute_fascicle_axon_density(axon_df, nerve_data, nerve_mask):
     """
-    Computes axon density per fascicle based on axon positions and fascicle areas.
+    Assigns axons to the labeled fascicles in 'nerve_mask', 
+    then uses the area for each label from 'nerve_data' to compute density.
     """
     fascicle_densities = {}
-    total_axons = len(axon_df)  
+    total_axons = len(axon_df)
 
-    fascicle_areas = {idx: fascicle["value"] for idx, fascicle in nerve_data["fascicle_areas"].items()}
-    
-    # extract (x, y) coordinates of axons
+    # read columns in (y, x) order so the loop matches (y, x)
     axon_positions = axon_df[["x0 (px)", "y0 (px)"]].values
 
-    # use KMeans clustering to assign axons to fascicles
-    # kmeans = KMeans(n_clusters=len(fascicle_areas), random_state=42, n_init=10)
-    # axon_df["fascicle_id"] = kmeans.fit_predict(axon_positions)
-
-    #TODO: instead of using KMeans to assign fascile_id, use this nerve_fascicles array which is an image with values 0, 1, 2, 3, ... for each fascicle
     axon_df["fascicle_id"] = -1
-    nerve_fascicles = measure.label(nerve_mask)
-    #TODO: check axon centroid is inside which fascicle area
-    pass
 
-    # compute number of axons per fascicle
+    clean_mask = cleanup_nerve_mask(nerve_mask>0, min_object_size=5000, max_hole_size=5000)
+    nerve_fascicles = measure.label(clean_mask)
+
+    height, width = nerve_fascicles.shape
+
+    # assign each axon to the label in nerve_fascicles
+    for i, (x, y) in enumerate(axon_positions):
+        y_int, x_int = int(y), int(x)
+        if 0 <= y_int < height and 0 <= x_int < width:
+            label_id = nerve_fascicles[y_int, x_int]
+            if label_id > 0:
+                axon_df.loc[i, "fascicle_id"] = label_id
```

Comment:
hm I think that's where the problem is. `skimage.measure` considers 0 as background, and 1 is the first object.

```suggestion
                axon_df.loc[i, "fascicle_id"] = label_id - 1
```

Reply:
I think you're right, skimage.measure.label starts at 1 for objects, so subtracting 1 here makes sense to align with zero-based indexing elsewhere. I'll test it to make sure it works.